File: flutter/lib/features/notes/domain/note_category.dart
class NoteCategory {
  final int id;
  final int position;
  final String name;
  final String color;
  final String icon;
  final bool isPersonal;

  const NoteCategory({
    required this.id,
    required this.position,
    required this.name,
    required this.color,
    required this.icon,
    required this.isPersonal,
  });

  factory NoteCategory.fromJson(Map<String, dynamic> json) {
    return NoteCategory(
      id: json['id'] as int,
      position: json['position'] as int? ?? 0,
      name: json['name'] as String,
      color: json['color'] as String? ?? '#0052CC',
      icon: json['icon'] as String? ?? '\u{1F4DD}',
      isPersonal: json['is_personal'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'color': color,
        'icon': icon,
        'is_personal': isPersonal,
      };
}

File: backend/app/models/note_category.py
from sqlalchemy import CheckConstraint, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column

from ..database import Base


class NoteCategory(Base):
    __tablename__ = "note_categories"
    __table_args__ = (
        CheckConstraint(
            "(is_personal = true AND user_id IS NOT NULL) "
            "OR (is_personal = false AND user_id IS NULL)",
            name="ck_note_category_scope_user",
        ),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    family_id: Mapped[int] = mapped_column(
        ForeignKey("families.id", ondelete="CASCADE"), index=True
    )
    user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=True
    )
    is_personal: Mapped[bool] = mapped_column(default=True, index=True)
    position: Mapped[int] = mapped_column(default=0, index=True)
    name: Mapped[str] = mapped_column(String(50))
    color: Mapped[str] = mapped_column(String(7), default="#0052CC")
    icon: Mapped[str] = mapped_column(String(10), default="📝")
